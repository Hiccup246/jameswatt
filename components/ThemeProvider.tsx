"use client";

import React from "react";

export const ThemeContext = React.createContext<{
  siteTheme: string | undefined;
  flipTheme: Function;
}>({ siteTheme: undefined, flipTheme: () => {} });

export function isDarkMode() {
  return document.documentElement.classList.contains("dark");
}

export function setDOMTheme({ darkMode }: { darkMode: boolean }) {
  if (!localStorage || !document) return;

  if (darkMode) {
    localStorage.theme = "dark";
    document.documentElement.classList.add("dark");
  } else {
    localStorage.theme = "light";
    document.documentElement.classList.remove("dark");
  }
}

export default function ThemeProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [siteTheme, setSiteTheme] = React.useState<string | undefined>();

  React.useEffect(() => {
    if (isDarkMode()) {
      setSiteTheme("dark");
    } else {
      setSiteTheme("light");
    }
  }, []);

  function flipTheme() {
    if (isDarkMode()) {
      setSiteTheme("light");
      setDOMTheme({ darkMode: false });
    } else {
      setSiteTheme("dark");
      setDOMTheme({ darkMode: true });
    }
  }

  return (
    <ThemeContext.Provider value={{ siteTheme, flipTheme }}>
      {children}
    </ThemeContext.Provider>
  );
}
