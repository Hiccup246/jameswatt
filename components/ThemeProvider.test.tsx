import renderer from "react-test-renderer";
import ThemeProvider, { setDOMTheme } from "./ThemeProvider";

describe("ThemeProvider", () => {
  it("correctly renders its children", () => {
    const tree = renderer
      .create(
        <ThemeProvider>
          <div>
            <p>Hello There</p>
          </div>
        </ThemeProvider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("setDOMTheme", () => {
  it("correctly sets localstorage and document body classes when darkMode is false", () => {
    Object.create(localStorage);
    Object.defineProperty(document, "documentElement", {
      configurable: true,
      value: document.createElement("document"),
    });

    setDOMTheme({ darkMode: false });

    expect(localStorage.theme).toBe("light");
    expect(document.documentElement.classList).not.toContain("dark");
  });

  it("correctly sets localstorage and document body classes when darkMode is true", () => {
    Object.create(localStorage);
    Object.defineProperty(document, "documentElement", {
      configurable: true,
      value: document.createElement("document"),
    });

    setDOMTheme({ darkMode: true });

    expect(localStorage.theme).toBe("dark");
    expect(document.documentElement.classList).toContain("dark");
  });
});
