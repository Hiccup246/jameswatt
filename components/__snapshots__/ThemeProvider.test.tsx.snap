// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider correctly renders its children 1`] = `
<div>
  <p>
    Hello There
  </p>
</div>
`;
